File: experiments/adrianco/experiment-5/runs/language=java_model=claude-opus-4-8_tooling=none/rep1/src/test/java/com/brasilsoccer/mcp/StatisticsTest.java
/*
 * ============================================================================
 * StatisticsTest - BDD scenarios for aggregated statistics (spec category 5)
 * ============================================================================
 * Context:
 *   Verifies league-wide aggregates (average goals per match, home/away/draw
 *   rates that sum to ~100%) and the "biggest wins" ranking, matching the spec's
 *   statistical-analysis examples.
 * ============================================================================
 */
package com.brasilsoccer.mcp;

import com.brasilsoccer.mcp.data.KnowledgeBase;
import com.brasilsoccer.mcp.model.Match;
import com.brasilsoccer.mcp.query.Results;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DisplayName("Feature: Statistical analysis")
class StatisticsTest {

    private final KnowledgeBase kb = TestData.kb();

    @Test
    @DisplayName("Given a season, When league stats computed, Then averages are plausible")
    void averageGoalsPlausible() {
        Results.LeagueStats s = kb.leagueStats("Brasileirao", 2019);
        assertTrue(s.matches() > 300, "a full season has hundreds of matches");
        assertTrue(s.avgGoalsPerMatch() > 1.5 && s.avgGoalsPerMatch() < 4.0,
                "avg goals/match should be realistic, was " + s.avgGoalsPerMatch());
    }

    @Test
    @DisplayName("Given league stats, When result rates summed, Then they total ~100%")
    void resultRatesSumToHundred() {
        Results.LeagueStats s = kb.leagueStats("Brasileirao", 2019);
        double sum = s.homeWinRate() + s.awayWinRate() + s.drawRate();
        assertTrue(Math.abs(sum - 100.0) < 0.5, "rates should sum to ~100, were " + sum);
        // Football has a well-known home advantage.
        assertTrue(s.homeWins() > s.awayWins(), "home wins should exceed away wins");
    }

    @Test
    @DisplayName("Given biggest-wins query, When ranked, Then margins are non-increasing")
    void biggestWinsRanked() {
        List<Match> wins = kb.biggestWins("Brasileirao", null, 10);
        assertFalse(wins.isEmpty());
        for (int i = 1; i < wins.size(); i++) {
            assertTrue(wins.get(i - 1).margin() >= wins.get(i).margin(),
                    "biggest wins must be ordered by margin descending");
        }
        assertTrue(wins.get(0).margin() >= 5, "the biggest Serie A win is a blowout");
    }
}

File: experiments/adrianco/experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=ATDD/rep1/test/acceptance/statistics_test.exs
defmodule BrazilianSoccer.Acceptance.StatisticsTest do
  @moduledoc """
  Category 5: Statistical Analysis. Goals-per-match averages, home/away win
  rates and the biggest victories, via the `competition_statistics` tool.
  """
  use ExUnit.Case, async: true

  alias BrazilianSoccer.Test.MCPClient

  test "average goals per match for a season is a plausible football figure" do
    result =
      MCPClient.call("competition_statistics", %{
        "competition" => "Brasileirão",
        "season" => 2019
      })

    assert result["matches"] == 380
    assert result["avg_goals_per_match"] > 1.5 and result["avg_goals_per_match"] < 4.0

    assert_in_delta result["avg_goals_per_match"],
                    result["total_goals"] / result["matches"],
                    0.01
  end

  test "win-rate proportions add up to one" do
    result =
      MCPClient.call("competition_statistics", %{
        "competition" => "Brasileirão",
        "season" => 2019
      })

    total =
      result["home_win_rate"] + result["away_win_rate"] + result["draw_rate"]

    assert_in_delta total, 1.0, 0.001
    assert result["home_win_rate"] > result["away_win_rate"]
  end

  test "the biggest wins are ordered by goal margin" do
    result =
      MCPClient.call("competition_statistics", %{
        "competition" => "Brasileirão",
        "season" => 2019,
        "biggest_wins_limit" => 5
      })

    wins = result["biggest_wins"]
    assert length(wins) == 5

    margins = Enum.map(wins, &abs(&1["home_goal"] - &1["away_goal"]))
    assert margins == Enum.sort(margins, :desc)
    # The single most lopsided result of the season had at least a 4-goal margin.
    assert hd(margins) >= 4
  end

  test "statistics can be aggregated across a whole competition (all seasons)" do
    result = MCPClient.call("competition_statistics", %{"competition" => "Brasileirão"})

    assert result["matches"] > 380
    assert result["avg_goals_per_match"] > 1.5 and result["avg_goals_per_match"] < 4.0
  end
end
